upgrade/v110to120/upgrade.go

```diff
@@ -260,124 +308,150 @@ func translateVolumeRecurringJobsToLabel(namespace string, lhClient *lhclientset
 		if err != nil {
 			return errors.Wrapf(err, "failed to list all cron jobs for volume %v", v.Name)
 		}
 		for name := range appliedCronJobROs {
-			log.Infof(upgradeLogPrefix+"Deleting %v cronjob job for %v volume", name, v.Name)
+			run.log.Infof(upgradeLogPrefix+"Deleting %v cronjob job for %v volume", name, v.Name)
 			err := cronJobClient.Delete(context.TODO(), name, metav1.DeleteOptions{PropagationPolicy: &propagation})
 			if err != nil {
 				return errors.Wrapf(err, "failed to delete %v cron job for volume %v", name, v.Name)
 			}
 		}
-		log.Infof(upgradeLogPrefix+"Deleted %v cronjob job for %v volume", len(appliedCronJobROs), v.Name)
+		run.log.Infof(upgradeLogPrefix+"Deleted %v cronjob job for %v volume", len(appliedCronJobROs), v.Name)
+	}
+
+	return nil
+}
+
+func (run *recurringJobUpgrade) createRecurringJobCRs() (err error) {
+	revertLog := run.log
+	defer func() {
+		run.log = revertLog
+		err = errors.Wrapf(err, "create recurringJob CRs failed")
+	}()
+	run.log = run.log.WithField("action", "create-recurringJob-CRs")
+
+	if len(run.recurringJobMapSpec) == 0 {
+		run.log.Debug(upgradeLogPrefix + "Found 0 recurring job")
+		return nil
+	}
+
+	recurringJobClient := run.lhClient.LonghornV1beta1().RecurringJobs(run.namespace)
+	for recurringJobName, spec := range run.recurringJobMapSpec {
+		run.log.Infof(upgradeLogPrefix+"Creating %v recurring job CR", recurringJobName)
+		newRecurringJob := &longhorn.RecurringJob{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: recurringJobName,
+			},
+			Spec: *spec,
+		}
+		run.log.Debugf(upgradeLogPrefix+"Checking if %v recurring job CR already exists", recurringJobName)
+		obj, err := recurringJobClient.Get(context.TODO(), newRecurringJob.Name, metav1.GetOptions{})
+		if err == nil {
+			run.log.Debugf(upgradeLogPrefix+"Recurring job CR already exists %v", obj)
+			continue
+		}
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to get recurring job %v", recurringJobName)
+		}
+		_, err = recurringJobClient.Create(context.TODO(), newRecurringJob, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "failed to create recurring job CR with %v", spec)
+		}
 	}
 	return nil
 }
 
-func buildStorageClassFromYAML(storageclassYAML string) (*storagev1.StorageClass, error) {
+func (run *recurringJobUpgrade) getStorageClassFromYAML(storageclassYAML string) error {
 	decode := scheme.Codecs.UniversalDeserializer().Decode
 	obj, _, err := decode([]byte(storageclassYAML), nil, nil)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to decoding YAML string")
+		return errors.Wrapf(err, "failed to decoding YAML string")
 	}
-	storageclass, ok := obj.(*storagev1.StorageClass)
+	sc, ok := obj.(*storagev1.StorageClass)
 	if !ok {
-		return nil, fmt.Errorf("invalid storageclass YAML string: %v", storageclassYAML)
+		return fmt.Errorf("invalid storageclass YAML string: %v", storageclassYAML)
 	}
-	return storageclass, nil
+	run.storageClass = sc
```

Comment:
The function naming is quite misleading because the property update happening in `getXXX` function.` I would suggest changing to other function name like `initStorageClassProFromYAML`.

Reply:
Addressed.